runtime/autoload/health/provider.vim

```diff
@@ -286,148 +251,74 @@ function! s:check_python(version) abort
+  if !empty(python_exe) && !empty(venv) 
+    \ && s:python_exepath(python_exe) =~# '^'.venv.'/'
+    \ && !exists('g:'.host_prog_var)
+    call health#report_warn('Your virtualenv is not set up optimally.', [
+          \ 'Set g:'.host_prog_var.' and install the pynvim module ONLY in that environment.',
+          \ 'You can use a global environment, or a virtualenv created specifically for Nvim.',
+          \ 'This will avoid the need to install the pynvim module in each virtualenv.'])
   endif
 
   " Diagnostic output
-  call health#report_info('Executable: ' . (empty(python_exe) ? 'Not found' : python_exe))
+  call health#report_info('Python provider executable: ' . (empty(python_exe) ? 'Not found' : python_exe))
+  let python_multiple = provider#pythonx#GetPythonCandidates(a:version, $PATH, v:true)
   if len(python_multiple)
+    " This is worth noting since the user may install something that changes
+    " $PATH, like Homebrew.
+    call health#report_info(printf('Multiple Python %d candidates found.  '
+          \ . 'Set g:%s to avoid surprises.', a:version, host_prog_var))
     for path_bin in python_multiple
-      call health#report_info('Other python executable: ' . path_bin)
+      if path_bin !=# python_exe
+        call health#report_info(printf('Other Python %d candidate: %s', a:version, path_bin))
+      endif
     endfor
   endif
 
-  let pip = 'pip' . (a:version == 2 ? '' : '3')
-
   if empty(python_exe)
     " No Python executable can import 'neovim'. Check if any Python executable
     " can import 'pynvim'. If so, that Python failed to import 'neovim' as
     " well, which is most probably due to a failed pip upgrade:
     " https://github.com/neovim/neovim/wiki/Following-HEAD#20181118
-    let [pynvim_exe, errors] = provider#pythonx#DetectByModule('pynvim', a:version)
+    let [pynvim_exe, _, errors] = provider#pythonx#DetectByModule('pynvim', a:version)
     if !empty(pynvim_exe)
+      let pip = pynvim_exe . ' -m pip'
       call health#report_error(
             \ 'Detected pip upgrade failure: Python executable can import "pynvim" but '
             \ . 'not "neovim": '. pynvim_exe,
             \ "Use that Python version to reinstall \"pynvim\" and optionally \"neovim\".\n"
             \ . pip ." uninstall pynvim neovim\n"
             \ . pip ." install pynvim\n"
-            \ . pip ." install neovim  # only if needed by third-party software")
+            \ . pip ." install neovim  # only if needed by third-party software\n"
+            \ . 'Retry with the --user flag if you get a permission error.')
     endif
   else
-    let [pyversion, current, latest, status] = s:version_info(python_exe)
-
-    if a:version != str2nr(pyversion)
-      call health#report_warn('Unexpected Python version.' .
-                  \ ' This could lead to confusing error messages.')
-    endif
-
-    if a:version == 3 && str2float(pyversion) < 3.3
-      call health#report_warn('Python 3.3+ is recommended.')
-    endif
```

Comment:
These are some of the happy-go-lucky version checks I mentioned; they are no longer needed, `pythonx/provider` makes sure it only returns an appropriate Python 2 or 3 version that satisfies minimum version requirements.